### ISSUE TITLE
`keyof` constrained generic is too wide

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#30993]
**TypeScript Version:**  3.4.1

**Search Terms:** generic constrained constraint keyof string number symbol PropertyKey

**Code**

```ts
{
    // declare const obj: { [P in string]: unknown }
    declare const obj: Record<string, unknown>
    declare const key: keyof typeof obj;
    // No error, good :-)
    const key2: string = key;
}

{
    const fn = <T extends Record<string, unknown>, K extends keyof T>(obj: T, key: K) => {
        /* Unexpected error:
        Type 'K' is not assignable to type 'string'.
            Type 'keyof T' is not assignable to type 'string'.
                Type 'string | number | symbol' is not assignable to type 'string'.
                    Type 'number' is not assignable to type 'string'. */
        const key2: string = key
    };
}
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:** 

**Related Issues:**

**Comment by OliverJAsh** [comment]
Also (potentially a separate issue):

``` ts
{
    declare const obj: { [key: string]: unknown };
    declare const key: keyof typeof obj;
    // Error, bad :-(
    const key2: string = key;
}
```

**Comment by jack-williams** [comment]
I think the error looks right here. The constraint on T just says: if a string key exists, it must have type unknown. The constraint never actually guarantees any keys, so it’s possible you pass something with symbol keys and a key that has type symbol.

**Comment by ahejlsberg** [comment]
The error is correct here as per @jack-williams above.

**Comment by Dudeonyx** [comment]
You can do

```ts
const fn = <T extends Record<string, unknown>, K extends keyof T>(obj: T, key: Extract<K, string>) => { 
//...
 }
```

**Comment by OliverJAsh** [comment]
That makes sense, thank you.

I guess my next question would be: how can I constrain the generic so that _all_ keys are `string`s (rather than _some_)?

**Comment by OliverJAsh** [comment]
> how can I constrain the generic so that _all_ keys are `string`s (rather than _some_)?

Maybe:

``` ts
{
    const fn = <K extends string, V, K2 extends K>(obj: Record<K, V>, key: K2) => {
        const key2: string = key
    };
}
```
